### ISSUE TITLE
[HOLD for payment 2025-01-02] [$250] Android - Category - "Default tax rate" briefly changes back to the default one

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#50078]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.43-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

**Issue found when executing PR** https://github.com/Expensify/App/pull/49867
## Action Performed:
1. Open the app
2. Log in with a new expensifail account
3. Create a workspace
4. Navigate to Workspace settings - More features
5. Enable Taxes, Rules and Workflows
4. Navigate to Workspace settings - Categories
5. Tap on any category
6. Tap on "Default tax rate"
7. Select "Tax Rate 1 - 5%"
8. Tap on "Name"
9. Add a few characters after the current name and tap on the "Save" button

## Expected Result:
"Default tax rate" should be the one I have set before

## Actual Result:
"Default tax rate" briefly changes back to the default one when changing category name for the first time

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/a9a66180-ca10-4649-979b-4be72ce8754e

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021844119049804579315</li>
        <li>Upwork Job ID: 1844119049804579315</li>
        <li>Last Price Increase: 2024-11-20</li>
<li>Automatic offers: </li>
<ul>
<li>allgandalf | Contributor | 105084228</li>
<li>wildan-m | Contributor | 105358083</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

"Default tax rate" briefly changes back to the default one

### What is the root cause of that problem?

The `defaultTaxRateText` still receives the old `categoryName` after returning from editing the category name
https://github.com/Expensify/App/blob/0c2900cb60e3e22ac5ca22e8485bb357de64db67/src/pages/workspace/categories/CategorySettingsPage.tsx#L84

And the `categoryName` is taken from the URL parameters, and the URL doesn't update `categoryName` right away. So, after navigating back, it still uses the old `categoryName`, causing this issue.
https://github.com/Expensify/App/blob/0c2900cb60e3e22ac5ca22e8485bb357de64db67/src/pages/workspace/categories/CategorySettingsPage.tsx#L38

### What changes do you think we should make in order to solve the problem?
We should set the params for navigation before navigating back, something like this:
```diff
// .src/pages/workspace/categories/EditCategoryPage.tsx#L47

    const editCategory = useCallback(
        (values: FormOnyxValues<typeof ONYXKEYS.FORMS.WORKSPACE_CATEGORY_FORM>) => {
            ....
+            Navigation.setParams({categoryName: newCategoryName});
            Navigation.goBack();
        },
        [backTo, currentCategoryName, route.params.categoryName, route.params.policyID],
    );
```
<summary>
<details>POC

https://github.com/user-attachments/assets/d8786a44-3dd7-4be1-af28-2df364968c5b

</details>
</summary>

**Comment by JmillsExpensify** [comment]
Opened up for proposal review

**Comment by JmillsExpensify** [comment]
Still waiting on proposal review.

**Comment by abdulrahuman5196** [comment]
Checking now

**Comment by abdulrahuman5196** [comment]
I don't see the issue anymore, can we check if its still reproducible?


https://github.com/user-attachments/assets/9abb6d85-c676-4421-a561-0ce5cb579446

**Comment by huult** [comment]
Still able to be reproduced
<details>
<summary>

https://github.com/user-attachments/assets/8a91b399-ef6a-49f2-9133-3658c67f7f40

</summary>
</details>

**Comment by huult** [comment]
## [Proposal updated](https://github.com/Expensify/App/issues/50078#issuecomment-2391824562)

- Change to a new solution.

**Comment by abdulrahuman5196** [comment]
@huult Do we need any more separate steps to repro the issue? 

Note: Anyways will check again as well on repro.

**Comment by huult** [comment]
@abdulrahuman5196 Can you create a new workspace and follow the steps in this ticket? This issue just happened once.

**Comment by huult** [comment]
## [Proposal updated](https://github.com/Expensify/App/issues/50078#issuecomment-2391824562)

- Update RCA
- Add alternative solutions

@abdulrahuman5196 I would like to share this video with you to show how we can reproduce this issue.

<details>

https://github.com/user-attachments/assets/ee9f6db8-a47a-4181-b60f-e73d270ee9c2

</details>

**Comment by abdulrahuman5196** [comment]
Thank you for this. Let me try again a couple of times today and update here.

**Comment by JmillsExpensify** [comment]
Still in progress

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Android: "Default tax rate" temporarily reverts to the default value.

### What is the root cause of that problem?

There is a chance that we navigated before execute Onyx.update first. The blinking actually also affect the category name itself, not only rate.

### What changes do you think we should make in order to solve the problem?

In this case, it's important to call `Onyx.update` before proceeding with the navigation action. We can achieve this by using `Navigation.setNavigationActionToMicrotaskQueue` as it's a commonly used pattern.

https://github.com/search?q=repo%3AExpensify%2FApp+Navigation.setNavigationActionToMicrotaskQueue%28&type=code

Modify this code:

https://github.com/Expensify/App/blob/d365d398b00077eac3eab8c9f812bd29df6a6823/src/pages/workspace/categories/EditCategoryPage.tsx#L52-L56

To

```tsx

            Navigation.setNavigationActionToMicrotaskQueue(()=>{
                Navigation.goBack(
                    isQuickSettingsFlow
                        ? ROUTES.SETTINGS_CATEGORY_SETTINGS.getRoute(route.params.policyID, route.params.categoryName, backTo)
                        : ROUTES.WORKSPACE_CATEGORY_SETTINGS.getRoute(route.params.policyID, route.params.categoryName),
                );
            });

```

or only to this if category name changes

```tsx

    const editCategory = useCallback(
        (values: FormOnyxValues<typeof ONYXKEYS.FORMS.WORKSPACE_CATEGORY_FORM>) => {
            const newCategoryName = values.categoryName.trim();
            const navigateBack = () => {
                Navigation.goBack(
                    isQuickSettingsFlow
                        ? ROUTES.SETTINGS_CATEGORY_SETTINGS.getRoute(route.params.policyID, route.params.categoryName, backTo)
                        : ROUTES.WORKSPACE_CATEGORY_SETTINGS.getRoute(route.params.policyID, route.params.categoryName),
                );
            };
    
            // Do not call the API if the edited category name is the same as the current category name
            if (currentCategoryName !== newCategoryName) {
                Category.renamePolicyCategory(route.params.policyID, {oldName: currentCategoryName, newName: values.categoryName});
                Navigation.setNavigationActionToMicrotaskQueue(navigateBack);
                return;
            }
    
            navigateBack();
        },
        [isQuickSettingsFlow, currentCategoryName, route.params.categoryName, route.params.policyID, backTo],
    );


```

[Branch for this solution](https://github.com/Expensify/App/compare/main...wildan-m:wildan/fix/50078-fix-old-data-briefly-appeared?expand=1)

### What alternative solutions did you explore? (Optional)
N/A

**Comment by JmillsExpensify** [comment]
@abdulrahuman5196 any thoughts on the latest proposal posted?

**Comment by abdulrahuman5196** [comment]
Hi, Checking now

**Comment by abdulrahuman5196** [comment]
Hi, I am still unable to repro the issue. I reached out in C+ channel to see, if any one is able to repro this and takeover.
https://expensify.slack.com/archives/C02NK2DQWUX/p1731523034411719

**Comment by wildan-m** [comment]
I consistently reproduce it by creating a new workspace for each test rather than reusing the same one.

**Comment by abdulrahuman5196** [comment]
@JmillsExpensify Kindly assign this issue to @allgandalf . Unassigning myself.

**Comment by JmillsExpensify** [comment]
Sorry missed this! Just assigned you.

**Comment by JmillsExpensify** [comment]
Still being worked on.

**Comment by wildan-m** [comment]
@allgandalf The PR is ready https://github.com/Expensify/App/pull/54239. Thanks!

**Comment by JmillsExpensify** [comment]
Apologies for the delay ya'll. Let me post the payment summary and get this one settled. 

Payment summary:

 - Contributor: @wildan-m $250 
 - Reviewer: @allgandalf $250

**Comment by JmillsExpensify** [comment]
All contributors paid via Upwork, and regression test created. I'm closing this out!
